models/DataObject/Concrete/Dao/InheritanceHelper.php

```diff
@@ -421,43 +421,64 @@ protected function buildTree($currentParentId, $fields = '', $parentIdGroups = n
 
         if (!$parentIdGroups) {
             $object = DataObject::getById($currentParentId);
-            if (isset($params['language'])) {
-                $query = "SELECT a.language as language, b.o_id AS id $fields, b.o_classId AS classId, b.o_parentId AS parentId FROM objects b LEFT JOIN " . $this->storetable . ' a ON b.o_id = a.' . $this->idField . ' WHERE o_path LIKE ' . $this->db->quote($object->getRealFullPath() . '/%')
-                    . ' HAVING `language` = "' . $params['language'] . '" OR ISNULL(`language`)'
-                    . ' ORDER BY LENGTH(o_path) ASC';
-            } else {
-                $query = "SELECT b.o_id AS id $fields, b.o_classId AS classId, b.o_parentId AS parentId FROM objects b LEFT JOIN " . $this->storetable . ' a ON b.o_id = a.' . $this->idField . ' WHERE o_path LIKE ' . $this->db->quote($object->getRealFullPath().'/%') . ' GROUP BY b.o_id ORDER BY LENGTH(o_path) ASC';
-            }
-            $queryCacheKey = 'tree_'.md5($query);
 
-            if (self::$useRuntimeCache) {
-                $parentIdGroups = self::$runtimeCache[$queryCacheKey] ?? null;
-            }
+            // we call hasChildren() here directly on the DAO to avoid the caching in the model
+            if($object->getDao()->hasChildren([$object::OBJECT_TYPE_OBJECT, $object::OBJECT_TYPE_FOLDER, $object::OBJECT_TYPE_VARIANT],true)){
+                $query = 'SELECT o_id FROM objects WHERE o_path LIKE ' .  $this->db->quote($object->getRealFullPath() . '/%');
+
+                if (self::$useRuntimeCache) {
+                    $queryCacheKey = 'tree_ids_'.md5($query);
+                    if(!key_exists($queryCacheKey,self::$runtimeCache)){
+                        self::$runtimeCache[$queryCacheKey] = $this->db->fetchCol($query);
+                    }
+                    $ids = self::$runtimeCache[$queryCacheKey];
+                }else{
+                    $ids = $this->db->fetchCol($query);
+                }
 
-            if (!$parentIdGroups) {
-                $result = $this->db->fetchAll($query);
 
                 if (isset($params['language'])) {
-                    $result = $this->filterResultByLanguage($result, $params['language'], 'language');
+                    $query = "SELECT a.language as language, b.o_id AS id $fields, b.o_classId AS classId, b.o_parentId AS parentId FROM objects b LEFT JOIN " . $this->storetable . ' a ON b.o_id = a.' . $this->idField . ' WHERE  o_id IN( ' . implode(',',$ids).')'
```

Comment:
I have a concern regarding the `o_id IN()` condition. If you have hundreds of thousands children, you're probably hitting `max_allowed_packet` .